### DISCUSSION TITLE
Asset management for our websites like lifeitself.org: separating from main site

### DISCUSSION THREAD

**Discussion by rufuspollock** [life-itself/community#436] [General]
This is a thread for discussing improvements to our asset management (images, pdfs etc) for lifeitself.org and other websites.

This follows up previous discussions of DAM (digital asset management) solutions plus trial with cloudinary in 2021.

# Image storage discussion

### Situation

- Currently store all images and other static assets directly into the main site folder on github (using git-lfs)
- We are storing lots of images for our ecosystem wiki (~100s or even 1000s). We don't want to think about the sizes of images. Images often end up being ~0.5-1MB => total amount could be GB or more.

### Complication

- slows build as have to download 1GB+ of assets
- working with repo is more painful as have
- no image resizing atm
- github won't like having GBs of images. Putting them in LFS is one solution but makes stuff more complex and painful e.g. you can't upload files directly on github interface (any other issues?).

### Question

where should we store images and how does that integrate with resizing such that ...

- ...
- ...

### Hypothesis

No answer yet!

Note pros:

- Assets and files in same place which means adding stuff is super simple (no need for two steps: upload my image, then link it etc)

What are the alternatives?

- Use cloudflare r2 with a custom domain plus image resizing - https://developers.cloudflare.com/r2/buckets/public-buckets/
- Cost?
- How to upload?
  - r2 dashboard or wrangler (needs access to cloudflare)
  - any r2 compatible command line tool with correct keys
  - (someday) custom plugin for obsidian like https://github.com/gavvvr/obsidian-imgur-plugin for imgur

## Notes

#todo dig out old issues about DAM etc

### What we would love

- integrated experience for editors
  - just paste images or other assets and they automatically get uploaded
  - widget for resizing (or good instructions for doing so in raw html etc)
- sustainability and simplicity of tech architecture (both on cost and team time)

### Obsidian plugins

- https://github.com/addozhang/obsidian-image-upload-toolkit
- https://github.com/gavvvr/obsidian-imgur-plugin ⭐175
- https://github.com/renmu123/obsidian-image-auto-upload-plugin ⭐286

**Comment by rufuspollock** [comment]
@nathenf @khalilcodes @olayway would love your thoughts here.

**Comment by olayway** [comment]
@rufuspollock I'd just use one of the available Obsidian plugins - or fork and slightly adjust first if needed.

There is not only the plugin for uploading to imgur. There is also one for Cloudinary - just tested it and it works great (although signing requests is not implemented). And there is another, generic one, where you can pass any API endpoint, headers, and body of your requests - so in theory, this can be used with any hosting provider. However, for uploading to e.g. R2 this would not work out of the box as the requests need to be signed, so we'd need to tweak the code a bit. But I think it's doable. Plus it also includes image resize settings.

Here are the two plugins:
- https://github.com/jordanhandy/obsidian-cloudinary-uploader
- https://github.com/Creling/obsidian-image-uploader

**Reply by rufuspollock** [reply]
@olayway cool. the main question would be indeed doing the signing etc. this is where you need both login for users and then that login to link with our signing microservice. we've built this ourselves a few times and i wonder if there is something off the shelf e.g. using supabase.

**Reply by rufuspollock** [reply]
I note supabase may soon have support for r2 as a backend (though not sure you can get that support from cloud version): https://github.com/supabase/storage-api/pull/350

**Comment by rufuspollock** [comment]
## Bytescale looks like a great semi-open-source option with backing onto r2 (from $35/month)

Just came across Bytescale: https://www.bytescale.com/pricing (formely upload.io)

Present as a developer oriented cloundinary alternative.

More relevantly for us they are a frontend and processing pipeline for r2 (or s3 etc) which gives us a combination of UX and portability (from our own storage).

![image](https://github.com/life-itself/community/assets/180658/15676968-68c1-4531-96ab-bdc430b25e7a)

Notes

- Added R2 support July 2023 https://www.bytescale.com/blog/cloudflare-r2-support-released/
- Need the $35/month plan (vs $7 basic) to support r2 and to have up to 5 team members (vs 1 on the basic plan)

### Good, fast, simple upload experience

here's the result

![image](https://github.com/life-itself/community/assets/180658/cd1e3dd7-2104-4bf5-8a7e-4f5f3c2c2575)
